src/Compilers/Core/Portable/Compilation/CompilationStuff.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Concurrent;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    /// <summary>
+    /// A mutable typesafe dictionary from a pair of compilation and key (of type <see cref="CompilationStuff.Key{T}"/>)
+    /// to a value (of type T). All keys and values associated with a compilation are reachable for GC purposes as a
+    /// consequence of this API for only as long as the compilation is reachable. These extension methods make a
+    /// compilation act as a property bag but without the <see cref="Compilation"/> itself being mutable.
+    /// </summary>
+    public static class CompilationStuff
+    {
+        public class Key<T> { }
```

Comment:
It's not supposed to be abstract. You are supposed to create (singleton) instances of it to use as keys for the "stuff":